stan/math/prim/prob/neg_binomial_lpmf.hpp

```diff
@@ -65,69 +65,46 @@ return_type_t<T_shape, T_inv_scale> neg_binomial_lpmf(const T_n& n,
     }
   }
 
-  VectorBuilder<true, T_partials_return, T_inv_scale> log_beta(size_beta);
+  VectorBuilder<true, T_partials_return, T_inv_scale> log1p_inv_beta(size_beta);
   VectorBuilder<true, T_partials_return, T_inv_scale> log1p_beta(size_beta);
-  VectorBuilder<true, T_partials_return, T_inv_scale> log_beta_m_log1p_beta(
-      size_beta);
   for (size_t i = 0; i < size_beta; ++i) {
     const T_partials_return beta_dbl = value_of(beta_vec[i]);
-    log_beta[i] = log(beta_dbl);
+    log1p_inv_beta[i] = log1p(inv(beta_dbl));
     log1p_beta[i] = log1p(beta_dbl);
-    log_beta_m_log1p_beta[i] = log_beta[i] - log1p_beta[i];
   }
 
-  VectorBuilder<true, T_partials_return, T_shape, T_inv_scale> lambda(
-      size_alpha_beta);
-  VectorBuilder<true, T_partials_return, T_shape, T_inv_scale>
-      alpha_log_beta_over_1p_beta(size_alpha_beta);
   VectorBuilder<!is_constant_all<T_inv_scale>::value, T_partials_return,
                 T_shape, T_inv_scale>
       lambda_m_alpha_over_1p_beta(size_alpha_beta);
-  for (size_t i = 0; i < size_alpha_beta; ++i) {
-    const T_partials_return alpha_dbl = value_of(alpha_vec[i]);
-    const T_partials_return beta_dbl = value_of(beta_vec[i]);
-    lambda[i] = alpha_dbl / beta_dbl;
-    alpha_log_beta_over_1p_beta[i] = alpha_dbl * log_beta_m_log1p_beta[i];
-    if (!is_constant_all<T_inv_scale>::value) {
-      lambda_m_alpha_over_1p_beta[i] = lambda[i] - alpha_dbl / (1 + beta_dbl);
+  if (!is_constant_all<T_inv_scale>::value) {
+    for (size_t i = 0; i < size_alpha_beta; ++i) {
+      const T_partials_return alpha_dbl = value_of(alpha_vec[i]);
+      const T_partials_return beta_dbl = value_of(beta_vec[i]);
```

Comment:
Have a look at what I did. I don't see this pattern elsewhere in the code, but I think I did the right thing.

Is this what you meant? If so I can bench is real quick since it doesn't look familiar but I agree it makes sense (we avoid a bunch of vari dereferences which are probably expensive)

Reply:
I did a benchmark for a length N input (where alpha and beta are both length N vectors). There wasn't much of a difference. It was worth a try but I'm going to revert the change just so this code looks like all the other lpmf stuff.

With the new code:

```
----------------------------------------------------------------------
Benchmark                               Time           CPU Iterations
----------------------------------------------------------------------
neg_binomial/2/manual_time            509 ns        544 ns    1368991
neg_binomial/4/manual_time            822 ns        855 ns     841446
neg_binomial/8/manual_time           1483 ns       1517 ns     473999
neg_binomial/16/manual_time          2797 ns       2830 ns     247871
neg_binomial/32/manual_time          5425 ns       5458 ns     129481
neg_binomial/64/manual_time         10607 ns      10637 ns      66323
neg_binomial/128/manual_time        21443 ns      21485 ns      33076
neg_binomial/256/manual_time        42787 ns      42817 ns      16309
neg_binomial/512/manual_time        85370 ns      85407 ns       8247
neg_binomial/1024/manual_time      170181 ns     170208 ns       4116
```

And then the old code:
```
neg_binomial2/2/manual_time           487 ns        522 ns    1435154
neg_binomial2/4/manual_time           806 ns        840 ns     859681
neg_binomial2/8/manual_time          1505 ns       1540 ns     469482
neg_binomial2/16/manual_time         2836 ns       2870 ns     247975
neg_binomial2/32/manual_time         5403 ns       5436 ns     128561
neg_binomial2/64/manual_time        10801 ns      10834 ns      65618
neg_binomial2/128/manual_time       21387 ns      21429 ns      32484
neg_binomial2/256/manual_time       42970 ns      43009 ns      16359
neg_binomial2/512/manual_time       85111 ns      85145 ns       8208
neg_binomial2/1024/manual_time     169757 ns     169782 ns       4128
```

This is the vague benchmark code:
```
static void neg_binomial(benchmark::State& state) {
  using stan::math::var;
  std::vector<int> y(state.range(0), 7);
  Eigen::VectorXd alpha_vals = 0.5 * Eigen::VectorXd::Ones(state.range(0)) + 0.1 * Eigen::VectorXd::Random(state.range(0));
  Eigen::VectorXd beta_vals = 0.5 * Eigen::VectorXd::Ones(state.range(0)) + 0.1 * Eigen::VectorXd::Random(state.range(0));
  for (auto _ : state) {
    auto start = std::chrono::high_resolution_clock::now();
    Eigen::Matrix<var, -1, 1> alpha = alpha_vals;
    Eigen::Matrix<var, -1, 1> beta = beta_vals;
    var lp = neg_binomial_lpmf(y, alpha, beta);
    lp.grad();
    auto end = std::chrono::high_resolution_clock::now();
    auto elapsed_seconds =
      std::chrono::duration_cast<std::chrono::duration<double>>(end - start);
    state.SetIterationTime(elapsed_seconds.count());
    stan::math::recover_memory();
    benchmark::ClobberMemory();
  }
}
```

Reply:
Neat, no differences, safe to ignore